Add NCS example to showcase LLPM features (#1344)
NordicPlayground/fw-nrfconnect-nrf: samples/bluetooth/llpm/README.rst

.. list-table:: GATT Attributes
   :header-rows: 1

   * - Type
     - UUID
     - Property
   * - Primary service
     - BT_UUID_LATENCY
     - Read only
   * - Characteristic
     - BT_UUID_LATENCY_CHAR
     - BT_GATT_CHRC_WRITE, BT_GATT_PERM_WRITE

This sample transmits data between two boards to measure the transmission latency in between.
One of the devices is connected as a *master* and another is connected as a *slave*.
The performance is evaluated with the transmission latency dividing the estimated round-trip time in half (RTT / 2).

By default, the following values are used to demonstrates the interaction of the connection parameters:

.. list-table:: Default parameter values
   :header-rows: 1

   * - Parameter
     - Value
   * - Connection interval
     - 80 units (100 ms)
   * - LLPM connection interval
     - Lowest interval (1 ms)
   * - Physical layer (PHY)
     - LE 2M PHY

Requirements
************

* Two of the following nRF52-series development kit boards:

  * PCA10040
  * PCA10056

>   * The others running LLPM-supported SoftDevice (see :ref:`nrfxlib:ble_controller` Proprietary feature support)

Author: Per discussion with @greg-fer, I will rephrase it to BLE Controller variants in the document.

.. _ble_llpm:

Bluetooth: LLPM
###############

> The Bluetooth Low Latency Packet Mode (LLPM) sample uses the :ref:`latency_readme` to showcase the LLPM proprietary Bluetooth extension from Nordic Semiconductor.

Reviewer: Would rephrase it to "The Bluetooth Low Latency Packet Mode (LLPM) sample uses the :ref:`throughput_readme` to measure the GATT Write latency. It uses the properietary LLPM Bluetooth extensions to achieve a latency below 7.5 ms. The sample can also be used to determine how different connection parameters affect the latency"

Reviewer 2: Agree - I was looking for an information that this is not a Bluetooth standard.

You can use it to determine the transmission latency of LLPM-enabled connections, or to compare with different connection parameters and check their influence on the results.

Overview
********

> The LLPM is designed time-sensitive transmissions, in which time synchronization is important and interface response time is critical for the user.

Reviewer: Would rephrase to: "LLPM is designed for applications where interface response time is critical for the user". Time synchronization is not one of the main use cases

> The LLPM is designed time-sensitive transmissions, in which time synchronization is important and interface response time is critical for the user.

Reviewer 3: ```suggestion
The LLPM is designed for applications in which the interface response time is critical for the user.
```

> For example, for reality headsets or gaming mouse and keyboard.

Author: @greg-fer How do you think?

> For example, for reality headsets or gaming mouse and keyboard.

Reviewer 4: I'm just asking. Maybe I just don't know what reality headset are 😛

> For example, for reality headsets or gaming mouse and keyboard.

Reviewer 3: ```suggestion
For example, for virtual reality headsets or gaming mouse and keyboard.
```
Do you mean virtual reality headsets? I understood this was the meaning.

> 

Reviewer: I would insert the following here:

```suggestion
This sample determines the latency as shown in the MSC below::
```
(and then the MSC which is now placed further down)
In this way we make it clear how the latency is defined at an early point in the documentation.

See the following subsections for a description of the key LLPM elements.

LLPM connection interval (1 ms)
   The connection interval defines how often the devices must listen on the radio.
   The LLPM introduces the possibility to reduce the connection interval below what is supported in BLE.

>    The lowest interval supported is 1 ms for one link.

Reviewer: "The lowest supported connection interval is 1 ms"

Reviewer 3: I agree. Let's rather have a link from the example to the service.

   The report gives information about the quality of service of the connection event.
   The values in the report are used to describe the quality of links.
   For parameter descriptions, see :cpp:enum:`hci_vs_evt_qos_conn_event_report_t` (in :file:`ble_controller_hci_vs.h`).

Transmission latency
   The definition of the latency used in this example counts the time interval from the sender's application to the GATT service of the receiver.
   It demonstrates the performance of an LLPM-enabled connection that the receiver will receive the data approximately every 1 ms.

GATT Latency Service
   To measure the transmission latency from application layers, a GATT Latency service `BT_UUID_LATENCY` is included to compute the time spent.
   When the sender writes its timestamp to the `BT_UUID_LATENC_CHAR` characteristic of the receiver, the Latency service of the receiver will automatically reply back.
   Whenever the sender receives a response, it will use its present time and the corresponding timestamp written before to estimate the round-trip time (RTT) of a writing characteristic procedure (see `Bluetooth Core Specification`_: Vol 3, Part F, 3.4.5 Writing attributes).

> .. list-table:: GATT Attributes

Reviewer: Should the service definition table for the latency service be moved to the latency service documentation? I believe it should be removed from the sample documentation

   :header-rows: 1

   * - Type
     - UUID
     - Property
   * - Primary service
     - BT_UUID_LATENCY
     - Read only
   * - Characteristic
     - BT_UUID_LATENCY_CHAR
     - BT_GATT_CHRC_WRITE, BT_GATT_PERM_WRITE

> This sample transmits data between two boards to measure the transmission latency in between.

Reviewer: In the sample documentation we mention 3 times how the latency is computed. I would say that keeping only the MSC is sufficient.

One of the devices is connected as a *master* and another is connected as a *slave*.
The performance is evaluated with the transmission latency dividing the estimated round-trip time in half (RTT / 2).

By default, the following values are used to demonstrates the interaction of the connection parameters:

.. list-table:: Default parameter values
   :header-rows: 1

   * - Parameter
     - Value
   * - Connection interval
     - 80 units (100 ms)
   * - LLPM connection interval
     - Lowest interval (1 ms)
   * - Physical layer (PHY)
     - LE 2M PHY

Requirements
************

* Two of the following nRF52-series development kit boards:

>   * PCA10040

Reviewer 3: The boards are so different that I don't see a way to rephrase this in a PCA-agnostic way. To my understanding, the samples will be checked every time for each new board, so the documentation will have to be updated at that stage. On the other hand, omitting this information from the sample documentation will create confusion.

>   * PCA10040

Reviewer: What about: 
Two nRF52-series development kit boards, e.g.:
?

>   * PCA10040

Author: How about linking to the README of ble-controller (nrfxlib\ble_controller\README.rst)?

+--------------------------+--------+--------+--------+
|                          | S112   | S132   | S140   |
+==========================+========+========+========+
| Low Latency Packet mode  |        | X      | X      |
+--------------------------+--------+--------+--------+
| QoS Conn Event Reports   | X      | X      | X      |
+--------------------------+--------+--------+--------+

>   * PCA10040

Reviewer 4: Leave as is. This is how it is done for all samples, so we should keep the pattern here as well.

  * PCA10056

  You can mix different boards.
* Connection to a computer with a serial terminal for each of the boards.

Building and running
********************
.. |sample path| replace:: :file:`samples/bluetooth/llpm`

.. include:: /includes/build_and_run.txt

Testing
=======

After programming the sample to both boards, test it by performing the following steps:

1. Connect to both boards with a terminal emulator (for example, PuTTY).
   See :ref:`putty` for the required settings.
#. Reset both boards.
#. Observe that the boards establish a connection.
   When they are connected, one of them serves as *master* and the other one as *slave*.

   - The master outputs the following information::

       Press any key to set LLPM short connection connection interval (1 ms)

   - The slave outputs the following information::

       Press any key to start measuring transmission latency

#. Press a key in the terminal that is connected to the slave.

> #. Observe that the slave starts sending data to the master and measuring the transmission latency with the default connection interval::

Reviewer: Would rephrase to:
```suggestion
#. Observe that latency measurements are printed in the terminal. The latency is expected to be slightly shorter than the connection interval:
```

Reviewer 3: As per RST formatting, the colon needs to be double `::`. Otherwise, the RST will treat the single colon as a linking markup and break the formatting.

       Transmission Latency: 80917 (us)

#. Press a key in the terminal that is connected to the master.
#. Observe the output on the master.
   It is waiting for a user input::

       Press any key to start measuring transmission latency

#. Observe the connection gets updated to LLPM connection interval (1 ms) on both sides::

>        Connection intervonal updated: LLPM (1 ms)

Reviewer: Typo:
```suggestion
       Connection interval updated: LLPM (1 ms)
```

#. Observe the output on the slave.

>    While the slave keeps sending data to the master with the LLPM connection interval, the measured latency on the slave becomes approximate 1 ms::

Reviewer: Suggested rephrasing: 
"On the terminal connected to the slave, observe the latency is reduced to approximately 1 ms."

       Transmission Latency: 1098 (us)

#. Press a key in the terminal that is connected to the master.

> #. Observe the output on the master.

Reviewer: I would simplify this phrasing to be similar to the slave.
"On the terminal connected to the master, observe the latency is approximately 1 ms."

   The master starts sending data to the slave and measuring the transmission latency with LLPM connection interval.
   The latency remains approximate 1 ms::

       Transmission Latency: 1235 (us)

.. msc::
   hscale = "1.3";
   Master,Slave;
   Master<<=>>Slave         [label="Connected"];
   Master<<=>>Slave         [label="Discovered BT_UUID_LATENCY SERVICE"];
   Slave note Slave         [label="Press any key to start measuring transmission latency"];

>    Slave note Slave         [label="Read present timestamp: s1"];

Reviewer: Should we use "Read current timestamp" instead of "Read present timestamp" instead? Similar for other occurrences in this MSC

   Slave=>Master            [label="ATT Write Request (timestamp: s1)"];
   Master>>Slave            [label="ATT Write response"];
   Slave note Slave         [label="Read present timestamp: s2"];

>    Slave note Slave         [label="Slave Latency = (s2 - s1) / 2"];

Reviewer: To avoid confusion, we should rather use the term "Latency, Slave" as "Slave Latency" is a BLE Spec feature. Similar for "Master Latency"

Author: To avoid confusion, maybe we should just use "Latency = ...". Because we only divide the RTT in half, it is hard to judge that the latency belongs to master or slave. ??

   Transmission Latency: 80093 (us)
   Transmission Latency: 80017 (us)
   Transmission Latency: 79940 (us)
   Transmission Latency: 79849 (us)
   Connection interval updated: LLPM (1 ms)
   Transmission Latency: 81604 (us)
   Transmission Latency: 30181 (us)
   Transmission Latency: 1098 (us)
   QoS conn event reports: channel index 0x22, CRC errors 0x00
   Transmission Latency: 1129 (us)
   Transmission Latency: 1037 (us)
   Transmission Latency: 930 (us)
   Transmission Latency: 1312 (us)
   Transmission Latency: 1083 (us)
   Transmission Latency: 1007 (us)

Dependencies
*************

This sample uses the following |NCS| libraries:

* :ref:`latency_readme`

In addition, it uses the following Zephyr libraries:

* :file:`include/console.h`
* :ref:`zephyr:kernel`:

  * :file:`include/kernel.h`

* :file:`include/misc/printk.h`
* :file:`include/zephyr/types.h`
* :ref:`zephyr:bluetooth_api`:

  * :file:`include/bluetooth/bluetooth.h`
  * :file:`include/bluetooth/conn.h`
  * :file:`include/bluetooth/gatt.h`
  * :file:`include/bluetooth/hci.h`
  * :file:`include/bluetooth/uuid.h`
  * :file:`include/bluetooth/scan.h`
  * :file:`include/bluetooth/gatt_dm.h`

> Moreover, it uses the following Nrfx libraries:

Reviewer: @greg-fer  , how do we document this properly? "It uses the BLE controller library in nrfxlib"?

Reviewer 3: ```suggestion
Moreover, it uses the following nrfxlib library::
```
I'd say it like this. The repetition doesn't hurt, and we keep consistency with the previous sentence about Zephyr libs.

Reviewer: Should it rather be moved right below "This sample uses the following |NCS| libraries:"

Reviewer 3: Need to check.

Reviewer 3: OK, checked. Please keep the nrfxlib line, but move it to line 277 (move the Zephyr sentence down). The nrfxlib sentence should read:
```
This sample uses the following `nrfxlib`_ libraries:

* :ref:`nrfxlib:ble_controller`
```
We don't need the file mentioned under this ref.

* :ref:`nrfxlib:ble_controller`:

  * :file:`include/ble_controller_hci_vs.h`

References
***********

For more information about the connection parameters that are used in this sample, see the following chapters of the `Bluetooth Core Specification`_:

> * Vol 3, Part F, 3.2.8 Exchanging MTU Size

Reviewer: ```suggestion
```

* Vol 3, Part F, 3.4.5 Writing attributes
* Vol 3, Part G, 4.9.3 Write Characteristic Value
* Vol 6, Part B, 5.1.1 Connection Update Procedure

> * Vol 6, Part B, 5.1.9 Data Length Update Procedure

Reviewer: ```suggestion

```
